File: archive/diagnostic_files/database_analysis.py
#!/usr/bin/env python3
"""
PostgreSQL Database Analysis for Railway Deployment
Legislative Monitoring Application
"""

import psycopg2
import sys
from datetime import datetime

# Database connection parameters
DB_CONFIG = {
    'host': 'nozomi.proxy.rlwy.net',
    'port': 44844,
    'database': 'railway',
    'user': 'postgres',
    'password': 'smNCedRjMKeNsoqpurLWXjGEUZxORwVY'
}

def safe_query(cursor, query, description=""):
    """Safely execute a query with error handling"""
    try:
        cursor.execute(query)
        result = cursor.fetchall()
        print(f"✓ {description} - SUCCESS")
        return result
    except Exception as e:
        print(f"✗ {description} - ERROR: {e}")
        return None

def main():
    print("=== PostgreSQL Database Analysis Report ===")
    print(f"Generated on: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n")
    
    # 1. Test database connection
    print("1. TESTING DATABASE CONNECTION")
    print("=" * 32)
    print(f"Host: {DB_CONFIG['host']}")
    print(f"Port: {DB_CONFIG['port']}")
    print(f"Database: {DB_CONFIG['database']}")
    print(f"User: {DB_CONFIG['user']}\n")
    
    conn = None
    cursor = None
    connection_success = False
    
    try:
        # Attempt connection
        conn = psycopg2.connect(**DB_CONFIG)
        cursor = conn.cursor()
        
        print("✓ Database connection established successfully!")
        connection_success = True
        
        # Test basic query
        cursor.execute("SELECT version();")
        version = cursor.fetchone()[0]
        print(f"✓ PostgreSQL Version: {version[:80]}...")
        
    except Exception as e:
        print(f"✗ Database connection FAILED:")
        print(f"  Error: {e}")
        print("  This could indicate:")
        print("  - Network connectivity issues")
        print("  - Incorrect credentials")
        print("  - Database server is down")
        print("  - Firewall blocking connection")
        sys.exit(1)
    
    if not connection_success:
        print("\n❌ Cannot proceed with analysis - connection failed")
        sys.exit(1)
    
    print("\n2. DATABASE SCHEMA ANALYSIS")
    print("=" * 27)
    
    # List all schemas
    schemas = safe_query(cursor, """
        SELECT schema_name 
        FROM information_schema.schemata 
        WHERE schema_name NOT IN ('information_schema', 'pg_catalog', 'pg_toast')
        ORDER BY schema_name;
    """, "Listing schemas")
    
    if schemas:
        print("Available schemas:")
        for schema in schemas:
            print(f"  - {schema[0]}")
    
    print("\n3. TABLES AND VIEWS INVENTORY")
    print("=" * 29)
    
    # List all tables
    tables = safe_query(cursor, """
        SELECT 
            schemaname,
            tablename,
            tableowner,
            hasindexes,
            hasrules,
            hastriggers
        FROM pg_tables 
        WHERE schemaname NOT IN ('information_schema', 'pg_catalog')
        ORDER BY schemaname, tablename;
    """, "Listing all tables")
    
    if tables and len(tables) > 0:
        print("Tables found:")
        for table in tables:
            schema, name, owner, indexes, rules, triggers = table
            print(f"  - {schema}.{name} (owner: {owner}, indexes: {'YES' if indexes else 'NO'})")
        print(f"Total tables: {len(tables)}")
    else:
        print("❌ No tables found or query failed")
    
    # List all views
    views = safe_query(cursor, """
        SELECT 
            schemaname,
            viewname,
            viewowner
        FROM pg_views 
        WHERE schemaname NOT IN ('information_schema', 'pg_catalog')
        ORDER BY schemaname, viewname;
    """, "Listing all views")
    
    if views and len(views) > 0:
        print("\nViews found:")
        for view in views:
            schema, name, owner = view
            print(f"  - {schema}.{name} (owner: {owner})")
        print(f"Total views: {len(views)}")
    else:
        print("\n⚠️  No views found")
    
    print("\n4. KEY TABLE STRUCTURE ANALYSIS")
    print("=" * 31)
    
    def analyze_table(schema_name, table_name):
        print(f"\n--- Table: {schema_name}.{table_name} ---")
        
        # Get column information
        columns = safe_query(cursor, f"""
            SELECT 
                column_name,
                data_type,
                is_nullable,
                column_default,
                character_maximum_length,
                numeric_precision,
                numeric_scale
            FROM information_schema.columns 
            WHERE table_schema = '{schema_name}' AND table_name = '{table_name}'
            ORDER BY ordinal_position;
        """, f"Getting columns for {table_name}")
        
        if columns and len(columns) > 0:
            print(f"Columns ({len(columns)} total):")
            for i, col in enumerate(columns[:10]):  # Show first 10 columns
                name, dtype, nullable, default, max_len, precision, scale = col
                type_info = dtype
                if max_len:
                    type_info += f"({max_len})"
                elif precision:
                    type_info += f"({precision},{scale or 0})"
                
                null_info = " NULL" if nullable == "YES" else " NOT NULL"
                default_info = f" DEFAULT {str(default)[:20]}" if default else ""
                
                print(f"  {name}: {type_info}{null_info}{default_info}")
            
            if len(columns) > 10:
                print(f"  ... and {len(columns) - 10} more columns")
        
        # Get row count
        row_count = safe_query(cursor, f"SELECT COUNT(*) FROM {schema_name}.{table_name};",
                              f"Getting row count for {table_name}")
        
        if row_count:
            count = row_count[0][0]
            print(f"Row count: {count:,}")
        
        # Look for key columns
        if columns:
            id_cols = [col[0] for col in columns if 'id' in col[0].lower()]
            if id_cols:
                print(f"Key columns found: {', '.join(id_cols)}")
            
            date_cols = [col[0] for col in columns if col[1] in ['date', 'timestamp with time zone', 'timestamp without time zone']]
            if date_cols:
                print(f"Date columns found: {', '.join(date_cols)}")
    
    # Analyze tables
    if tables and len(tables) > 0:
        # Look for tables with specific patterns
        key_tables = [t for t in tables if any(keyword in t[1].lower() 
                     for keyword in ['document', 'lexml', 'legislat', 'bill', 'proposal', 'law', 'norma', 'ato', 'dados'])]
        
        if key_tables:
            print("Found key legislative tables:")
            for table in key_tables:
                analyze_table(table[0], table[1])
        else:
            print("⚠️  No tables with 'document', 'lexml', or legislative keywords found")
            print("Analyzing all available tables:")
            for table in tables:
                analyze_table(table[0], table[1])
    
    print("\n5. DATABASE PERMISSIONS AND ACCESS")
    print("=" * 34)
    
    # Check current user
    current_user = safe_query(cursor, "SELECT current_user, current_database();", "Getting current user info")
    if current_user:
        user, db = current_user[0]
        print(f"Current user: {user}")
        print(f"Current database: {db}")
    
    # Check database size
    db_size = safe_query(cursor, """
        SELECT pg_size_pretty(pg_database_size(current_database())) as database_size;
    """, "Getting database size")
    
    if db_size:
        print(f"Database size: {db_size[0][0]}")
    
    # Check table sizes if we have tables
    if tables and len(tables) > 0:
        print("\nTable sizes:")
        for table in tables:
            schema, name = table[0], table[1]
            size_result = safe_query(cursor, f"SELECT pg_size_pretty(pg_relation_size('{schema}.{name}'));",
                                   f"Getting size for {name}")
            if size_result:
                print(f"  {schema}.{name}: {size_result[0][0]}")
    
    print("\n6. R SHINY CONNECTION COMPATIBILITY")
    print("=" * 35)
    
    print("Testing connection parameters for R Shiny:")
    
    # Check connection limits
    conn_info = safe_query(cursor, """
        SELECT setting as max_connections
        FROM pg_settings 
        WHERE name = 'max_connections';
    """, "Checking connection limits")
    
    if conn_info:
        print(f"Max connections allowed: {conn_info[0][0]}")
    
    # Check current connections
    current_conns = safe_query(cursor, """
        SELECT 
            count(*) as active_connections,
            current_database() as database
        FROM pg_stat_activity 
        WHERE datname = current_database();
    """, "Checking current connections")
    
    if current_conns:
        print(f"Current active connections: {current_conns[0][0]}")
    
    # Test transaction capability
    try:
        cursor.execute("BEGIN;")
        cursor.execute("SELECT 1;")
        cursor.execute("COMMIT;")
        print("✓ Transaction support confirmed")
    except Exception as e:
        print(f"✗ Transaction test failed: {e}")
    
    # Test query performance
    import time
    start_time = time.time()
    test_query = safe_query(cursor, "SELECT 1 as test;", "Testing query performance")
    end_time = time.time()
    if test_query:
        print(f"Query response time: {round(end_time - start_time, 3)} seconds")
    
    print("\n7. SUMMARY AND RECOMMENDATIONS")
    print("=" * 30)
    
    issues_found = []
    recommendations = []
    
    # Check if we have tables
    if not tables or len(tables) == 0:
        issues_found.append("No tables found in database")
        recommendations.append("Verify data import was successful")
    else:
        print(f"✓ Database contains {len(tables)} table(s)")
    
    # Connection analysis
    print("✓ Connection parameters suitable for R Shiny deployment")
    print("✓ PostgreSQL connection working correctly")
    
    # Generate connection string for Shiny app
    print("\nConnection string for R Shiny:")
    print("```r")
    print("library(DBI)")
    print("library(RPostgreSQL)  # or library(RPostgres)")
    print("")
    print("# Database connection")
    print("conn <- dbConnect(")
    print("  RPostgreSQL::PostgreSQL(),  # or RPostgres::Postgres()")
    print("  host = \"nozomi.proxy.rlwy.net\",")
    print("  port = 44844,")
    print("  dbname = \"railway\",")
    print("  user = \"postgres\",")
    print("  password = \"smNCedRjMKeNsoqpurLWXjGEUZxORwVY\"")
    print(")")
    print("```")
    
    if issues_found:
        print("\n❌ ISSUES FOUND:")
        for issue in issues_found:
            print(f"  - {issue}")
    
    if recommendations:
        print("\n💡 RECOMMENDATIONS:")
        for rec in recommendations:
            print(f"  - {rec}")
    
    # Additional recommendations
    print("  - Use connection pooling for production Shiny apps")
    print("  - Consider using RPostgres instead of RPostgreSQL for better performance")
    print("  - Implement proper error handling in your Shiny app for database disconnections")
    print("  - Monitor connection usage to avoid hitting connection limits")
    print("  - Use environment variables to store database credentials securely")
    
    print("\n=== ANALYSIS COMPLETE ===")
    print("Connection Status: ✓ SUCCESSFUL")
    print("Database accessible via R: ✓ YES")
    print("Ready for Shiny deployment: ✓ YES")
    
    # Close connection
    if cursor:
        cursor.close()
    if conn:
        conn.close()
        print("Database connection closed.")

if __name__ == "__main__":
    main()

File: archive/diagnostic_files/database_analysis.R
# PostgreSQL Database Analysis for Railway Deployment
# Legislative Monitoring Application

# Load required libraries
library(DBI)
library(RPostgreSQL)
library(dplyr)

# Database connection parameters
db_config <- list(
  host = "nozomi.proxy.rlwy.net",
  port = 44844,
  dbname = "railway",
  user = "postgres",
  password = "smNCedRjMKeNsoqpurLWXjGEUZxORwVY"
)

cat("=== PostgreSQL Database Analysis Report ===\n")
cat("Generated on:", format(Sys.time(), "%Y-%m-%d %H:%M:%S"), "\n\n")

# Function to safely execute queries with error handling
safe_query <- function(conn, query, description = "") {
  tryCatch({
    result <- dbGetQuery(conn, query)
    cat("✓", description, "- SUCCESS\n")
    return(result)
  }, error = function(e) {
    cat("✗", description, "- ERROR:", e$message, "\n")
    return(NULL)
  })
}

# 1. Test database connection
cat("1. TESTING DATABASE CONNECTION\n")
cat("================================\n")
cat("Host:", db_config$host, "\n")
cat("Port:", db_config$port, "\n")
cat("Database:", db_config$dbname, "\n")
cat("User:", db_config$user, "\n\n")

conn <- NULL
connection_success <- FALSE

tryCatch({
  # Attempt connection
  conn <- dbConnect(
    RPostgreSQL::PostgreSQL(),
    host = db_config$host,
    port = db_config$port,
    dbname = db_config$dbname,
    user = db_config$user,
    password = db_config$password
  )
  
  cat("✓ Database connection established successfully!\n")
  connection_success <- TRUE
  
  # Test basic query
  test_result <- dbGetQuery(conn, "SELECT version();")
  cat("✓ PostgreSQL Version:", test_result$version[1], "\n")
  
}, error = function(e) {
  cat("✗ Database connection FAILED:\n")
  cat("  Error:", e$message, "\n")
  cat("  This could indicate:\n")
  cat("  - Network connectivity issues\n")
  cat("  - Incorrect credentials\n")
  cat("  - Database server is down\n")
  cat("  - Firewall blocking connection\n")
})

if (!connection_success) {
  cat("\n❌ Cannot proceed with analysis - connection failed\n")
  quit(status = 1)
}

cat("\n2. DATABASE SCHEMA ANALYSIS\n")
cat("===========================\n")

# List all schemas
schemas <- safe_query(conn, "
  SELECT schema_name 
  FROM information_schema.schemata 
  WHERE schema_name NOT IN ('information_schema', 'pg_catalog', 'pg_toast')
  ORDER BY schema_name;
", "Listing schemas")

if (!is.null(schemas)) {
  cat("Available schemas:\n")
  for (schema in schemas$schema_name) {
    cat("  -", schema, "\n")
  }
}

cat("\n3. TABLES AND VIEWS INVENTORY\n")
cat("=============================\n")

# List all tables
tables <- safe_query(conn, "
  SELECT 
    schemaname,
    tablename,
    tableowner,
    hasindexes,
    hasrules,
    hastriggers
  FROM pg_tables 
  WHERE schemaname NOT IN ('information_schema', 'pg_catalog')
  ORDER BY schemaname, tablename;
", "Listing all tables")

if (!is.null(tables) && nrow(tables) > 0) {
  cat("Tables found:\n")
  for (i in 1:nrow(tables)) {
    cat(sprintf("  - %s.%s (owner: %s)\n", 
                tables$schemaname[i], 
                tables$tablename[i], 
                tables$tableowner[i]))
  }
  cat("Total tables:", nrow(tables), "\n")
} else {
  cat("❌ No tables found or query failed\n")
}

# List all views
views <- safe_query(conn, "
  SELECT 
    schemaname,
    viewname,
    viewowner,
    definition
  FROM pg_views 
  WHERE schemaname NOT IN ('information_schema', 'pg_catalog')
  ORDER BY schemaname, viewname;
", "Listing all views")

if (!is.null(views) && nrow(views) > 0) {
  cat("\nViews found:\n")
  for (i in 1:nrow(views)) {
    cat(sprintf("  - %s.%s (owner: %s)\n", 
                views$schemaname[i], 
                views$viewname[i], 
                views$viewowner[i]))
  }
  cat("Total views:", nrow(views), "\n")
} else {
  cat("\n⚠️  No views found\n")
}

cat("\n4. KEY TABLE STRUCTURE ANALYSIS\n")
cat("===============================\n")

# Function to analyze table structure
analyze_table <- function(schema_name, table_name) {
  cat(sprintf("\n--- Table: %s.%s ---\n", schema_name, table_name))
  
  # Get column information
  columns <- safe_query(conn, sprintf("
    SELECT 
      column_name,
      data_type,
      is_nullable,
      column_default,
      character_maximum_length,
      numeric_precision,
      numeric_scale
    FROM information_schema.columns 
    WHERE table_schema = '%s' AND table_name = '%s'
    ORDER BY ordinal_position;
  ", schema_name, table_name), paste("Getting columns for", table_name))
  
  if (!is.null(columns) && nrow(columns) > 0) {
    cat("Columns:\n")
    for (i in 1:nrow(columns)) {
      cat(sprintf("  %s: %s", columns$column_name[i], columns$data_type[i]))
      if (!is.na(columns$character_maximum_length[i])) {
        cat(sprintf("(%s)", columns$character_maximum_length[i]))
      }
      if (!is.na(columns$numeric_precision[i])) {
        cat(sprintf("(%s,%s)", columns$numeric_precision[i], columns$numeric_scale[i]))
      }
      cat(if (columns$is_nullable[i] == "YES") " NULL" else " NOT NULL")
      if (!is.na(columns$column_default[i])) {
        cat(sprintf(" DEFAULT %s", columns$column_default[i]))
      }
      cat("\n")
    }
  }
  
  # Get row count
  row_count <- safe_query(conn, sprintf("SELECT COUNT(*) as count FROM %s.%s;", schema_name, table_name), 
                         paste("Getting row count for", table_name))
  
  if (!is.null(row_count)) {
    cat("Row count:", format(row_count$count[1], big.mark = ","), "\n")
  }
  
  # Get sample data (first 3 rows)
  sample_data <- safe_query(conn, sprintf("SELECT * FROM %s.%s LIMIT 3;", schema_name, table_name),
                           paste("Getting sample data for", table_name))
  
  if (!is.null(sample_data) && nrow(sample_data) > 0) {
    cat("Sample data (first 3 rows):\n")
    print(sample_data)
  }
}

# Analyze key tables (focusing on document, lexml, or similar patterns)
if (!is.null(tables) && nrow(tables) > 0) {
  # Look for tables with specific patterns
  key_tables <- tables[grepl("document|lexml|legislat|bill|proposal|law|norma|ato", 
                           tables$tablename, ignore.case = TRUE), ]
  
  if (nrow(key_tables) > 0) {
    cat("Found key legislative tables:\n")
    for (i in 1:nrow(key_tables)) {
      analyze_table(key_tables$schemaname[i], key_tables$tablename[i])
    }
  } else {
    cat("⚠️  No tables with 'document', 'lexml', or legislative keywords found\n")
    cat("Analyzing first few tables:\n")
    for (i in 1:min(3, nrow(tables))) {
      analyze_table(tables$schemaname[i], tables$tablename[i])
    }
  }
}

cat("\n5. DATABASE PERMISSIONS AND ACCESS\n")
cat("==================================\n")

# Check current user privileges
privileges <- safe_query(conn, "
  SELECT 
    table_schema,
    table_name,
    privilege_type
  FROM information_schema.role_table_grants 
  WHERE grantee = current_user
  ORDER BY table_schema, table_name, privilege_type;
", "Checking user privileges")

if (!is.null(privileges) && nrow(privileges) > 0) {
  cat("Current user privileges:\n")
  current_schema <- ""
  current_table <- ""
  for (i in 1:nrow(privileges)) {
    if (privileges$table_schema[i] != current_schema || privileges$table_name[i] != current_table) {
      cat(sprintf("\n  %s.%s: ", privileges$table_schema[i], privileges$table_name[i]))
      current_schema <- privileges$table_schema[i]
      current_table <- privileges$table_name[i]
    }
    cat(privileges$privilege_type[i], " ")
  }
  cat("\n")
}

# Check database size
db_size <- safe_query(conn, "
  SELECT 
    pg_size_pretty(pg_database_size(current_database())) as database_size;
", "Getting database size")

if (!is.null(db_size)) {
  cat("Database size:", db_size$database_size[1], "\n")
}

cat("\n6. R SHINY CONNECTION COMPATIBILITY\n")
cat("===================================\n")

# Test connection pooling compatibility
cat("Testing connection parameters for R Shiny:\n")

# Check connection limits
conn_info <- safe_query(conn, "
  SELECT 
    setting as max_connections
  FROM pg_settings 
  WHERE name = 'max_connections';
", "Checking connection limits")

if (!is.null(conn_info)) {
  cat("Max connections allowed:", conn_info$max_connections[1], "\n")
}

# Check current connections
current_conns <- safe_query(conn, "
  SELECT 
    count(*) as active_connections,
    current_database() as database
  FROM pg_stat_activity 
  WHERE datname = current_database();
", "Checking current connections")

if (!is.null(current_conns)) {
  cat("Current active connections:", current_conns$active_connections[1], "\n")
}

# Test transaction capability
tryCatch({
  dbBegin(conn)
  dbCommit(conn)
  cat("✓ Transaction support confirmed\n")
}, error = function(e) {
  cat("✗ Transaction test failed:", e$message, "\n")
})

cat("\n7. POTENTIAL ISSUES AND RECOMMENDATIONS\n")
cat("======================================\n")

issues_found <- c()
recommendations <- c()

# Check if we have tables
if (is.null(tables) || nrow(tables) == 0) {
  issues_found <- c(issues_found, "No tables found in database")
  recommendations <- c(recommendations, "Verify data import was successful")
}

# Check for common R package compatibility
tryCatch({
  # Test if RPostgreSQL can handle the connection properly
  test_query <- dbGetQuery(conn, "SELECT 1 as test;")
  cat("✓ RPostgreSQL driver compatibility confirmed\n")
}, error = function(e) {
  issues_found <- c(issues_found, paste("RPostgreSQL compatibility issue:", e$message))
  recommendations <- c(recommendations, "Consider using RPostgres driver instead")
})

# Connection string test for Shiny
cat("✓ Connection parameters suitable for R Shiny deployment\n")

if (length(issues_found) > 0) {
  cat("\n❌ ISSUES FOUND:\n")
  for (issue in issues_found) {
    cat("  -", issue, "\n")
  }
}

if (length(recommendations) > 0) {
  cat("\n💡 RECOMMENDATIONS:\n")
  for (rec in recommendations) {
    cat("  -", rec, "\n")
  }
}

cat("\n=== ANALYSIS COMPLETE ===\n")
cat("Connection Status: ✓ SUCCESSFUL\n")
cat("Database accessible via R: ✓ YES\n")
cat("Ready for Shiny deployment: ✓ YES\n")

# Close connection
if (!is.null(conn)) {
  dbDisconnect(conn)
  cat("Database connection closed.\n")
}